locator/tablets.cc

```diff
@@ -537,14 +537,18 @@ void tablet_aware_replication_strategy::validate_tablet_options(const gms::featu
     }
 }
 
-void tablet_aware_replication_strategy::process_tablet_options(abstract_replication_strategy& ars,
+void tablet_aware_replication_strategy::process_tablet_options(const gms::feature_service& fs,
+                                                               abstract_replication_strategy& ars,
                                                                replication_strategy_config_options& opts) {
-    auto i = opts.find(cql3::statements::ks_prop_defs::INITIAL_TABLETS_KEY);
-    if (i != opts.end()) {
-        _initial_tablets = parse_initial_tablets(i->second);
+    if (fs.tablets) {
```

Comment:
I think we should record this decision in the schema somehow to be safe.

Otherwise, during upgrade, one node may still think the feature is disabled and anther node may already see that the feature is enabled, and they'll disagree about whether the keyspace is using tablets.

Reply:
> I think we should record this decision in the schema somehow to be safe.

Would record(s) in system.tablets work?

> Otherwise, during upgrade, one node may still think the feature is disabled and anther node may already see that the feature is enabled, and they'll disagree about whether the keyspace is using tablets.

If a node tries to join the cluster with tablets enabled it will need to enable tablets too and not turn it off since then, so presumably you're talking about some early start-up code? Where can this misunderstanding currently happen?

Reply:
> > I think we should record this decision in the schema somehow to be safe.
> 
> Would record(s) in system.tablets work?

No, because partitions there are per-table not per keyspace.

> 
> > Otherwise, during upgrade, one node may still think the feature is disabled and anther node may already see that the feature is enabled, and they'll disagree about whether the keyspace is using tablets.
> 
> If a node tries to join the cluster with tablets enabled it will need to enable tablets too and not turn it off since then, so presumably you're talking about some early start-up code? Where can this misunderstanding currently happen?

I'm thinking about rolling upgrade which will enable tablets feature. There is a period where all nodes support the feature but not all of them enabled it.

Reply:
Actually current bootstrap code is already racy in this sense. Here's what I see from logs (some lines skipped):

```
16:24:05,391 storage_service - Performing gossip shadow round
16:24:05,391 gossip - Gossip shadow round started with nodes={127.83.115.13, 127.83.115.11, 127.83.115.7}
16:24:05,392 gossip - Gossip shadow round finished with nodes_talked={127.83.115.13, 127.83.115.7, 127.83.115.11}
16:24:05,392 storage_service - Checking bootstrapping/leaving/moving nodes: ok (check_for_endpoint_collision)
16:24:05,392 storage_service - Save advertised features list in the 'system.local' table
16:24:05,395 schema_tables - Schema version changed to 59adb24e-f3cd-3e02-97f0-5b395827453f
16:24:05,395 storage_service - Starting up server gossip
16:24:05,396 gossip - failure_detector_loop: Started main loop
16:24:05,396 raft_group0 - setup_group0: joining group 0...
16:24:05,396 raft_group0 - server c60209cb-28d4-4743-b59a-9e8602196367 found no local group 0. Discovering...
16:24:05,396 raft_group0 - server c60209cb-28d4-4743-b59a-9e8602196367 found group 0 with group id df10a1c0-99ba-11ee-9bf5-c955f5c66b4e, leader c80d3edd-fe43-41b6-ae33-8930afa41a4a
16:24:05,396 storage_service - raft topology: join: sending the join request to 127.83.115.11
16:24:05,415 storage_service - raft topology: join: request to join placed, waiting for the response from the topology coordinator
16:24:05,416 raft_group0 - Server c60209cb-28d4-4743-b59a-9e8602196367 is starting group 0 with id df10a1c0-99ba-11ee-9bf5-c955f5c66b4e
16:24:06,464 schema_tables - Creating keyspace system_auth
16:24:06,464 schema_tables - Creating keyspace system_distributed
16:24:06,464 schema_tables - Creating keyspace system_distributed_everywhere
16:24:06,464 schema_tables - Creating keyspace system_traces
16:24:06,464 schema_tables - Creating keyspace test
... <more "Creating keyspace ..." messages>
16:24:06,473 schema_tables - Tablet metadata changed
16:24:06,483 migration_manager - Gossiping my schema version e2c39868-99ba-11ee-47eb-96ff97452461
16:24:06,483 schema_tables - Schema version changed to e2c39868-99ba-11ee-47eb-96ff97452461
16:24:06,488 features - Feature TABLE_DIGEST_INSENSITIVE_TO_EXPIRY is enabled
...
16:24:06,535 features - Feature TABLETS is enabled
```

Note the sub-sequence:

```
16:24:06,464 schema_tables - Creating keyspace test
16:24:06,535 features - Feature TABLETS is enabled
```

So features are enabled in parallel (?) with merging schema and not seeing a feature when creating keyspace locally is becoming fatal.

Reply:
It used to be the case that gossip enabled features, so it was in the background. Now it's enabled through raft, so it should be easy to serialize it with schema changes and make sure it's committed before bootstrap completes. \cc @piodul How hard is that?

Reply:
Features are enabled in several iterations

```
INFO  2023-12-13 17:53:21,445 [shard 0:strm] features - Enabling feature list [{}] from topology state load
INFO  2023-12-13 17:53:21,445 [shard 0:strm] features - Enabling feature list finished
INFO  2023-12-13 17:53:22,502 [shard 0:stmt] schema_tables - do_merge_schema -> start
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_auth
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_distributed
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_distributed_everywhere
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_traces
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace test
... <more "Creating ..." messages>
INFO  2023-12-13 17:53:22,520 [shard 0:stmt] schema_tables - Tablet metadata changed
INFO  2023-12-13 17:53:22,527 [shard 0:stmt] schema_tables - do_merge_schema <- done
INFO  2023-12-13 17:53:22,539 [shard 0:strm] features - Enabling feature list [{AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECOND>
...
INFO  2023-12-13 17:53:22,591 [shard 0:strm] features - Feature TABLETS is enabled
...
INFO  2023-12-13 17:53:22,591 [shard 0:strm] features - Enabling feature list finished
```

the "Enabling feature list ... from topology state load" is the call to feature_service::enable() from storage_service::topology_state_load()

Reply:
Both feature-enabling happen from topology_sate_load() too and the "enabled_features" list is read from system.topology. 1st time it's empty, 2nd time it's full

Reply:
... and the update does _not_ happen via `topology_mutation_builder::add_enabled_features()`

Reply:
Features appear in system.topology while doing `transfer_snapshot()` that brings new schema to the node. So the steps are

* load_snapshot()
   - load_topology_state()
      - enable_feature( empty )
* transfer_snapshot()
   - merge_schema
      - create keyspaces
   - merge_topology_snapshot()
      - update system.topology.enabled_features
* load_snapshot() (another one?)
   - load_topology_state()
      - enable_feature( full )

Reply:
> So features are enabled in parallel (?) with merging schema and not seeing a feature when creating keyspace locally is becoming fatal.

@xemul In the logs you attached, it looks like a group0 snapshot transfer can be seen. In the snapshot transfer code we merge the schema first and only then call topology_state_load (where we mark features as enabled in feature_service).

\cc @piodul How hard is that?

@tgrabiec Both enabling a feature and modifying schema happens via a group0 command. If they are applied one after another then you are safe, but you may get a snapshot transfer. Perhaps we could change it to apply schema after features are marked as enabled, but @gleb-cloudius will most likely not like introducing such dependencies (my first design of cluster feature on raft did something like that and was rejected).

Besides, I see some other reasons not to do it like that:

- It's not clear to me whether there will be an issue in recovery. In recovery, we don't use raft to manage schema or cluster features.
- Relying on the property that "feature is enabled on this node => other node will see it as enabled at the right time" requires some non-local reasoning and knowing that the code is always called from group0 command application / snapshot transfer. As you mentioned, in general feature being enabled locally only implies that other nodes _support it_, but implies nothing about being seen as enabled.

Reply:
> @xemul In the logs you attached, it looks like a group0 snapshot transfer can be seen. In the snapshot transfer code we merge the schema first and only then call topology_state_load (where we mark features as enabled in feature_service).

Yes, I already [discovered](https://github.com/scylladb/scylladb/pull/16364#discussion_r1425664697) that

Reply:
> * load_snapshot()
>   
>   * load_topology_state()
>     
>     * enable_feature( empty )
> * transfer_snapshot()
>   
>   * merge_schema
>     
>     * create keyspaces
>   * merge_topology_snapshot()
>     
>     * update system.topology.enabled_features
> * load_snapshot() (another one?)
>   
>   * load_topology_state()
>     
>     * enable_feature( full )

It seems wrong that features are enabled post snapshot transfer (esp. schema), since snapshot content may depend on those features.

Reply:
> It seems wrong that features are enabled post snapshot transfer (esp. schema), since snapshot content may depend on those features.

The node suppose to be able to handle all features is supports even if they are not enabled yet. It only starts producing output that depends on the feature when it becomes enabled. This is the assumption with features over gossiper as well. Since the snapshot contains the feature as enabled the nodes that receives it has to support it already, so everything is fine.

Reply:
> Since the snapshot contains the feature as enabled the nodes that receives it has to support it already, so everything is fine.

Node does _support_ it already, the problem is that code than wants to see the tablets feature as _enabled_ is run first, then enabled features from the snapshot are enabled in the feature service

Reply:
> > Since the snapshot contains the feature as enabled the nodes that receives it has to support it already, so everything is fine.
> 
> Node does _support_ it already, the problem is that code than wants to see the tablets feature as _enabled_ is run first, then enabled features from the snapshot are enabled in the feature service

Why does the code cares if the feature is enabled in _this_ node? If it gets something tablet related from the other node it should be able to process it since this can only happen if the node supports the feature. It should not do `if (tablets not enabled yet) do something bad`. This is the same with gossiper IIRC. There is no guaranty all nodes see feature entanglements at the same time, what is guarantied is that if one node starts using the feature all nodes support it.

Reply:
> Why does the code cares if the feature is enabled in this node?

Because when we check features we do it like `if ((bool)feature_service.local().feature_name) { /* do something when feature is enabled */ } In particular, when a new node joins the cluster it merges schema, when schema is merged keyspaces are created, when a keyspace is created it creates replication strategy and if r.s. is the network replication strategy, its constructor checks for feature_service.tablets to be true/false (i.e. -- the feature is enabled/disabled) and then sets "per-table" bit accordingly. Not seeing the feature enabled early enough, results in vnode-based network replication strategy which is wrong, since all other nodes already see this keyspace as per-table r.s.

Reply:
>  its constructor checks for feature_service.tablets to be true/false (i.e. -- the feature is enabled/disabled) and then sets "per-table" bit accordingly.

I looks like the "per-table" bit  should be part of the schema, not decided locally based on incomplete info.

Reply:
> I looks like the "per-table" bit should be part of the schema, not decided locally based on incomplete info.

OK. Any place you have in mind? Part of `with replication = { ... }` set doesn't work, all field in it [are treated as replication factor integers](https://github.com/scylladb/scylladb/pull/16364#discussion_r1422569803)

Reply:
> > I looks like the "per-table" bit should be part of the schema, not decided locally based on incomplete info.
> 
> OK. Any place you have in mind? Part of `with replication = { ... }` set doesn't work, all field in it [are treated as replication factor integers](https://github.com/scylladb/scylladb/pull/16364#discussion_r1422569803)

We have scylla specific keyspace for our extensions: "scylla_keyspaces". IIRC there was a similar problem to what you describe in the link when we added in_memory option to tables (describe table failed)  and we solved by putting the option into scylla specific "scylla_tables".

Reply:
@tgrabiec , @gleb-cloudius , FYI, keeping this bit in system_schema.scylla_keyspaces steps on the same issue -- when merging schema this place

```c++
future<lw_shared_ptr<query::result_set>> extract_scylla_specific_keyspace_info(distributed<service::storage_proxy>& proxy, const schema_result_value_type& partition) {
    lw_shared_ptr<query::result_set> scylla_specific_rs;
    if (proxy.local().features().cluster_schema_features().contains<schema_feature::SCYLLA_KEYSPACES>()) {
        auto&& rs = partition.second;
        if (rs->empty()) {
            co_await coroutine::return_exception(std::runtime_error("query result has no rows"));
        }
        auto&& row = rs->row(0);
        auto keyspace_name = row.get_nonnull<sstring>("keyspace_name");
        auto keyspace_key = dht::decorate_key(*scylla_keyspaces(), partition_key::from_singular(*scylla_keyspaces(), keyspace_name));
        scylla_specific_rs = co_await db::system_keyspace::query(proxy.local().get_db(), NAME, SCYLLA_KEYSPACES, keyspace_key);
    }
    co_return scylla_specific_rs;
}
```

doesn't load anything from scylla_keyspaces because feature is not enabled yet

Reply:
Perhaps it should always extract (supported = may be there, enabled = allow putting it in).